File: app/core/axis_codes.py
from __future__ import annotations

from typing import Any, Iterable


IMPACT_AXIS_CODES: tuple[str, ...] = ("x+", "x-", "y+", "z+", "y-", "z-")
VIBRATION_AXIS_CODES: tuple[str, ...] = ("x", "y+", "z+", "y-", "z-")
DEFAULT_AXIS_CODES = IMPACT_AXIS_CODES
_AXIS_ORDER = {
    axis_code: index
    for index, axis_code in enumerate(("x", *IMPACT_AXIS_CODES))
}
_IMPACT_EXPERIMENT_TYPES = frozenset(("冲击试验", "冲击实验"))
_VIBRATION_EXPERIMENT_TYPES = frozenset(("振动试验", "振动实验"))


def normalize_text(value: Any) -> str:
    return str(value or "").strip()


def canonical_axis_code(value: Any) -> str:
    normalized = normalize_text(value)
    lowered = normalized.lower()
    return lowered if lowered in _AXIS_ORDER else normalized


def axis_codes_for_experiment_type(experiment_type: Any) -> tuple[str, ...]:
    normalized_type = normalize_text(experiment_type)
    if normalized_type in _IMPACT_EXPERIMENT_TYPES:
        return IMPACT_AXIS_CODES
    if normalized_type in _VIBRATION_EXPERIMENT_TYPES:
        return VIBRATION_AXIS_CODES
    return ()


def normalize_axis_codes_for_experiment_type(
    axis_codes: Iterable[Any],
    experiment_type: Any,
) -> list[str]:
    allowed_axis_codes = axis_codes_for_experiment_type(experiment_type)
    if not allowed_axis_codes:
        return []
    normalized: list[str] = []
    for item in axis_codes:
        axis_code = canonical_axis_code(item)
        if axis_code in allowed_axis_codes:
            normalized.append(axis_code)
    return [axis_code for axis_code in allowed_axis_codes if axis_code in set(normalized)]


def sort_axis_codes(axis_codes: Iterable[Any]) -> list[str]:
    ordered: list[tuple[int, str]] = []
    seen: set[str] = set()
    for index, item in enumerate(axis_codes):
        axis_code = canonical_axis_code(item)
        dedupe_key = axis_code.lower()
        if not axis_code or dedupe_key in seen:
            continue
        seen.add(dedupe_key)
        ordered.append((index, axis_code))
    ordered.sort(key=lambda item: (_AXIS_ORDER.get(item[1].lower(), len(_AXIS_ORDER)), item[0]))
    return [axis_code for _, axis_code in ordered]

File: frontend/src/lib/axisCodes.js
const IMPACT_AXIS_CODES = Object.freeze(["x+", "x-", "y+", "z+", "y-", "z-"]);
const VIBRATION_AXIS_CODES = Object.freeze(["x", "y+", "z+", "y-", "z-"]);
const DEFAULT_AXIS_CODES = IMPACT_AXIS_CODES;
const AXIS_ORDER_CODES = Object.freeze(["x", ...IMPACT_AXIS_CODES]);
const AXIS_ORDER = new Map(AXIS_ORDER_CODES.map((axisCode, index) => [axisCode, index]));
const IMPACT_EXPERIMENT_TYPES = new Set(["冲击试验", "冲击实验"]);
const VIBRATION_EXPERIMENT_TYPES = new Set(["振动试验", "振动实验"]);

const normalizeText = (value) => String(value ?? "").trim();

const canonicalAxisCode = (value) => {
  const normalized = normalizeText(value);
  const lowered = normalized.toLowerCase();
  return AXIS_ORDER.has(lowered) ? lowered : normalized;
};

const sortAxisCodes = (axisCodes = []) => {
  const seen = new Set();
  return (Array.isArray(axisCodes) ? axisCodes : [])
    .map((axisCode, index) => ({ axisCode: canonicalAxisCode(axisCode), index }))
    .filter(({ axisCode }) => {
      const dedupeKey = axisCode.toLowerCase();
      if (!axisCode || seen.has(dedupeKey)) {
        return false;
      }
      seen.add(dedupeKey);
      return true;
    })
    .sort((left, right) => {
      const leftOrder = AXIS_ORDER.get(left.axisCode.toLowerCase()) ?? AXIS_ORDER_CODES.length;
      const rightOrder = AXIS_ORDER.get(right.axisCode.toLowerCase()) ?? AXIS_ORDER_CODES.length;
      return leftOrder - rightOrder || left.index - right.index;
    })
    .map(({ axisCode }) => axisCode);
};

const parseAxisCodeInput = (value) => {
  if (Array.isArray(value)) {
    return value;
  }
  const normalized = normalizeText(value);
  if (!normalized) {
    return [];
  }
  try {
    const parsed = JSON.parse(normalized);
    if (Array.isArray(parsed)) {
      return parsed;
    }
  } catch {
    // Legacy text values use delimiters instead of JSON arrays.
  }
  return normalized.split(/[,，/、\s]+/);
};

const normalizeAxisCodes = (value) => sortAxisCodes(parseAxisCodeInput(value));

const axisCodesForExperimentType = (experimentType) => {
  const normalizedType = normalizeText(experimentType);
  if (IMPACT_EXPERIMENT_TYPES.has(normalizedType)) {
    return IMPACT_AXIS_CODES;
  }
  if (VIBRATION_EXPERIMENT_TYPES.has(normalizedType)) {
    return VIBRATION_AXIS_CODES;
  }
  return Object.freeze([]);
};

const normalizeAxisCodesForExperimentType = (value, experimentType) => {
  const allowedAxisCodes = axisCodesForExperimentType(experimentType);
  if (allowedAxisCodes.length === 0) {
    return [];
  }
  const selected = new Set(normalizeAxisCodes(value));
  return allowedAxisCodes.filter((axisCode) => selected.has(axisCode));
};

const formatAxisCodeLabel = (value) => normalizeText(value).toUpperCase();

export {
  DEFAULT_AXIS_CODES,
  IMPACT_AXIS_CODES,
  VIBRATION_AXIS_CODES,
  axisCodesForExperimentType,
  canonicalAxisCode,
  formatAxisCodeLabel,
  normalizeAxisCodes,
  normalizeAxisCodesForExperimentType,
  sortAxisCodes,
};
